### PR TITLE
Day 2 (2021)

### PR DESCRIPTION
**Pull request by nikola-rados**
Solves both part for [Advent of Code - Day 2](https://adventofcode.com/2021/day/2).

### CODE REVIEW

File: 2021/classes/submarine.py [hunk 1]
```diff
@@ -2,8 +2,9 @@
 
 
 class Submarine:
-    def __init__(self, readings = []):
+    def __init__(self, readings=[], move_commands=[]):
```
Comment (jgayfer) [inline 2021/classes/submarine.py]: This is a pretty Python specific piece of advice, but using an empty array as a default argument is potentially quite dangerous.

Little bit of reading here: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Instead, it's better to have a default of `None` and then check the value in `__init__` and set your instance variable to an empty list if desired.

File: 2021/classes/submarine.py [hunk 2]
```diff
@@ -17,3 +18,30 @@ def deep_sonar_sweep(self):
             for r1, r2, r3 in zip(self.readings, self.readings[1:], self.readings[2:])
         ]
         return self._sonar_sweep(triple_step_sums)
+
+    def simple_position(self):
+        horizontal, depth = 0, 0
+
+        for direction, distance in self.move_commands:
```
Comment (jgayfer) [inline 2021/classes/submarine.py]: The biggest improvement I would suggest here is to introduce a more strongly modeled "move command".

Right now we're using a tuple which is _fine_, and it works, but generally I find that any time you are passing around a tuple or a dictionary, there is very likely another object hidden away that we could take advantage of.

Consider if a move command gained an additional vector, such that we now had something like `(direction, distance, time)`. In order to incorporate this change, we'd need to change every iterator over `move_commands`, as there's now an additional member to unpack.

Now, consider how we'd handle that same situation if we had a more strongly modelled move command.

```python
class MoveCommand:
    # I'm using type hints here
    def __init__(self, direction: str, distance: int):
        self.direction = direction
        self.distance = distance
```

Now instead of changing our iterators can look like this:
```python
for move_command in self.move_commands:
    ...
```

And when we add a new attribute (such as a `time`), our iterator doesn't need to change!

In a simple example like this, it's not a _huge_ deal, but in larger code bases, extracting out data into strongly modelled objects can often make it easier to make changes easier later on.

**TL;DR** - More objects is _almost always_ a good direction to go. It's very tough to have too many objects in a system. I've found that some of the nicest systems to work in are the ones that have many small objects that composed together make up the system as a whole.

Comment (jgayfer) [reply 2021/classes/submarine.py]: As a bit of a bonus, we could also extract `direction` into an `Enum`, as then we can operate on strongly typed definitions, rather than naked strings, which is always nice, as it has the same side effect of making change easier to push through the system later on.

File: 2021/classes/submarine.py [hunk 3]
```diff
@@ -17,3 +18,30 @@ def deep_sonar_sweep(self):
...
+            if direction == "down":
+                depth += distance
+            elif direction == "up":
+                depth -= distance
+            elif direction == "forward":
+                horizontal += distance
+
+        return horizontal * depth
```
Comment (jgayfer) [inline 2021/classes/submarine.py:34]: 👍

File: 2021/classes/submarine.py [hunk 4]
```diff
@@ -17,3 +18,30 @@ def deep_sonar_sweep(self):
...
+        for direction, distance in self.move_commands:
+            if direction == "down":
+                aim += distance
+            elif direction == "up":
+                aim -= distance
+            elif direction == "forward":
+                horizontal += distance
+                depth += aim * distance
+
+        return horizontal * depth
```
Comment (jgayfer) [inline 2021/classes/submarine.py:48]: 👍

File: 2021/tests/test_submarine.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    "test_move_commands, expected",
+    [
+        ([("forward", 5), ("up", 2), ("down", 3)], 0),
+        ([("up", 5), ("up", 1), ("down", 10), ("forward", 5)], 100),
+        ([("down", 5), ("up", 2), ("forward", 10), ("up", 2), ("forward", 2)], 384),
+    ],
+)
+def test_complex_position(test_move_commands, expected):
+    sub = Submarine(move_commands=test_move_commands)
+    assert sub.complex_position() == expected
```
Comment (jgayfer) [inline 2021/tests/test_submarine.py:95]: Awesome!

**Comment by nikola-rados** [comment]
@jgayfer thanks for the feedback as always. I think I've addressed most of it, only item I'm a little uncertain of is the `enum` usage. Also you mentioned `pyright`, I tried installing it but it has some issues running alongside `pylance` which I guess already implements `pyright`. Would you recommend removing `pylance` in favour of just using `pyright`? I'm new to typing in python so I'm not sure what is best.

**Comment by jgayfer** [comment]
@nikola-rados If pylance works, I wouldn't be too worried! I think they more or less do the same thing.

The main benefit of the enum is that we now have a properly typed object. If you are given value of type `str` and are told that it can be one of three values, you have to more or less assume that the information that you have is correct, and it is only going to be one of those values.

Whereas if I give you a typed enum, that can only have three values, no matter what, I no longer need to explicitly tell you what values you need to expect, as they're built into the enum itself.

In the context of advent, it's not hugely obvious what the advantage of an enum is, but in a larger system where you and I might be building separate systems that need to interact with each other, removing any ambiguity of the API contract we both need to adhere to is a big win.
